<a href="https://colab.research.google.com/github/jorgevictor/Probabilidade-Estatistica/blob/main/ProbEst_Trab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Questao 1: Num jogo de craps dois dados são lançados. Se o jogador tira 7 ou 11 na soma das faces voltadas
para cima, ele ganha. Se ele tira 2, 3 ou 12 ele perde. Nos outros casos ele continua jogando os dois dados até
sair 7, caso em que ele perde, ou então sair novamente o primeiro resultado, caso em que ele ganha.
1. Escreva um programa de computador que simule um resultado de um jogo de Craps. O programa deve
retornar uma lista com um dos elementos sendo o vetor com os valores das somas dos dados obtidos em
sequência até terminar o jogo e o outro elemento sendo identificando se o resultado foi uma vitória ou uma
derrota (pode ser um inteiro 0 ou 1, ou então uma string, como preferir).
2. Rode o programa anterior 5 vezes e reporte os diferentes resultados obtidos (vetores com valores das faces
e a conclusão - vitória ou derrota).
3. Utilize o programa criado por você para estimar a probabilidade de vitória no jogo de Craps. Para isso
você deve rodar o programa muitas vezes (quanto mais vezes melhor) e calcular a proporção de partidas
que resultou em vitória.
4. Qual é a chance de que o jogo nunca termine? Justifique.
5. Se você paga 1 real quando perde o jogo e recebe 1 real como prêmio quando ganha, qual foi o saldo final
em reais ao final de todos os jogos simulados por você? O que se pode conjecturar sobre o resultado final
de muitas apostas sucessivas num jogo de Craps quanto ao retorno financeiro?

In [ ]:
import random

soma = 0
jogando = True
resultados = []
vitoria_derrota = []

while jogando:
    dado1 = random.randint(1, 6)
    dado2 = random.randint(1, 6)
    soma_atual = dado1 + dado2
    
    resultados.append(soma_atual)
    
    if soma == 0: # primeiro lance
        if soma_atual in [7, 11]:
            jogando = False
            vitoria_derrota.append("Vitória")
        elif soma_atual in [2, 3, 12]:
            jogando = False
            vitoria_derrota.append("Derrota")
        else:
            soma = soma_atual
            vitoria_derrota.append("Jogando")
    else: # já houve um lance anterior
        if soma_atual == 7:
            jogando = False
            vitoria_derrota.append("Derrota")
        elif soma_atual == soma:
            jogando = False
            vitoria_derrota.append("Vitória")
        else:
            vitoria_derrota.append("Jogando")

print("Resultados dos lances: ", resultados)
for i in range(len(vitoria_derrota)):
    print("Rodada {}: {}".format(i+1, vitoria_derrota[i]))


Resultados dos lances:  [5, 7]
Rodada 1: Jogando
Rodada 2: Derrota


In [ ]:
import random

n = 500000 # número de partidas a serem jogadas
vitorias = 0 # contador de vitórias

for i in range(n):
    soma = 0
    jogando = True
    resultados = []
    vitoria_derrota = []

    while jogando:
        dado1 = random.randint(1, 6)
        dado2 = random.randint(1, 6)
        soma_atual = dado1 + dado2

        resultados.append(soma_atual)

        if soma == 0: # primeiro lance
            if soma_atual in [7, 11]:
                jogando = False
                vitoria_derrota.append("Vitória")
            elif soma_atual in [2, 3, 12]:
                jogando = False
                vitoria_derrota.append("Derrota")
            else:
                soma = soma_atual
        else: # já houve um lance anterior
            if soma_atual == 7:
                jogando = False
                vitoria_derrota.append("Derrota")
            elif soma_atual == soma:
                jogando = False
                vitoria_derrota.append("Vitória")

    if "Vitória" in vitoria_derrota:
        vitorias += 1

print("Estimativa de vitórias em {} partidas: {} ({:.2f}%)".format(n, vitorias, vitorias/n * 100))


Estimativa de vitórias em 500000 partidas: 246241 (49.25%)


1.2)
Resultados dos lances:  [9, 8, 9]
Rodada 1: Jogando
Rodada 2: Jogando
Rodada 3: Vitória

Resultados dos lances:  [9, 7]
Rodada 1: Jogando
Rodada 2: Derrota

Resultados dos lances:  [6, 10, 2, 5, 7]
Rodada 1: Jogando
Rodada 2: Jogando
Rodada 3: Jogando
Rodada 4: Jogando
Rodada 5: Derrota

Resultados dos lances:  [8, 10, 11, 7]
Rodada 1: Jogando
Rodada 2: Jogando
Rodada 3: Jogando
Rodada 4: Derrota

Resultados dos lances:  [5, 7]
Rodada 1: Jogando
Rodada 2: Derrota

1.3) Estimativa de vitórias em 500000 partidas: 246241 (49.25%)

Ao rodar esse programa muitas vezes, é possível perceber que a estimativa sempre tende a 49.3%.

1.4) A probabilidade de que um jogo de craps nunca termine é extremamente baixa, na verdade, é praticamente impossível. A probabilidade de um jogo de craps terminar é sempre de 100%, pois as regras do jogo são projetadas para que o jogo tenha um fim.

O resultado de cada lançamento de dados em um jogo de craps é determinado pela probabilidade. Existem várias combinações possíveis de resultados, cada uma com sua própria probabilidade. Por exemplo, a probabilidade de se obter um total de 7 em dois dados é de 1/6, enquanto a probabilidade de se obter um total de 2 ou 12 é de apenas 1/36.

Independentemente do resultado de cada lançamento, o jogo continuará a progredir de acordo com as regras estabelecidas até que ocorra uma das seguintes situações: o jogador ganha e recolhe seus ganhos, o jogador perde e sai do jogo, ou o cassino decide encerrar o jogo.

Em resumo, a probabilidade de que um jogo de craps nunca termine é zero, pois as regras do jogo e as probabilidades envolvidas garantem que o jogo sempre chegará a uma conclusão.

1.5) Ao jogar mais e mais vezes, o saldo tende a ficar negativo por conta da probabilidade de vitória ser menor que 50%. Então, o que se pode conjecturar sobre o resultado final de muitas apostas sucessivas num jogo de Craps quanto ao retorno financeiro é que o retorno financeiro é negativo.

Questão 2: Em aula, calculamos a probabilidade de acertar exatamente 5 números na mega sena (acertar na
quina) tendo escolhido um bilhete de 6 números. Lembre-se que na Mega Sena são sorteados 6 números de 1 a
60 aleatoriamente sem repetições.
1. Escolha um bilhete de 6 números. Escreva um programa que simule um sorteio da Mega Sena e diga se
você acertou ou não na quina.
2. Utilize o programa criado por você para estimar a probabilidade de acertar na quina e compare com o
valor obtido em sala de aula.
3. Modifique o programa para verificar se você acertou exatamente 4 números ou não.
4. Utilize o programa criado no item anterior para estimar a probabilidade de acertar exatamente 4 números
da Mega Sena tendo escolhido um bilhete de 6 números.

In [ ]:
import random

# Definindo os números do bilhete
meu_bilhete = [1, 7, 13, 22, 36, 45]

# Gerando os números sorteados
numeros_sorteados = []
while len(numeros_sorteados) < 6:
    numero = random.randint(1, 60)
    if numero not in numeros_sorteados:
        numeros_sorteados.append(numero)

# Ordenando os números sorteados
numeros_sorteados.sort()

# Imprimindo o resultado
print("Seus números:", meu_bilhete)
print("Números sorteados:", numeros_sorteados)

# Verificando se acertou a quina
numeros_certos = 0
for numero in meu_bilhete:
    if numero in numeros_sorteados:
        numeros_certos += 1

if numeros_certos == 5:
    print("Parabéns, você acertou a quina!")
else:
    print("Infelizmente você não ganhou desta vez.")


Seus números: [1, 7, 13, 22, 36, 45]
Números sorteados: [20, 24, 36, 42, 48, 55]
Infelizmente você não ganhou desta vez.


In [ ]:
import random

n = 500000  # número de jogos a serem simulados
meu_bilhete = [1, 7, 13, 22, 36, 45]  # números escolhidos pelo jogador

num_quinas = 0  # contador de acertos da quina

for i in range(n):
    numeros_sorteados = []
    while len(numeros_sorteados) < 6:
        numero = random.randint(1, 60)
        if numero not in numeros_sorteados:
            numeros_sorteados.append(numero)
    numeros_sorteados.sort()

    numeros_certos = 0
    for numero in meu_bilhete:
        if numero in numeros_sorteados:
            numeros_certos += 1

    if numeros_certos == 5:
        num_quinas += 1

print("Probabilidade de acertar a quina em um jogo:", num_quinas/n)

Probabilidade de acertar a quina em um jogo: 4e-06


In [ ]:
import random

# Definindo os números do bilhete
meu_bilhete = [1, 7, 13, 22, 36, 45]

# Gerando os números sorteados
numeros_sorteados = []
while len(numeros_sorteados) < 6:
    numero = random.randint(1, 60)
    if numero not in numeros_sorteados:
        numeros_sorteados.append(numero)

# Ordenando os números sorteados
numeros_sorteados.sort()

# Imprimindo o resultado
print("Seus números:", meu_bilhete)
print("Números sorteados:", numeros_sorteados)

# Verificando se acertou a quina
numeros_certos = 0
for numero in meu_bilhete:
    if numero in numeros_sorteados:
        numeros_certos += 1

if numeros_certos == 4:
    print("Parabéns, você acertou exatamente 4 números!")
else:
    print("Infelizmente você não ganhou desta vez.")

Seus números: [1, 7, 13, 22, 36, 45]
Números sorteados: [8, 12, 47, 49, 54, 56]
Infelizmente você não ganhou desta vez.


In [ ]:
import random

n = 500000  # número de jogos a serem simulados
meu_bilhete = [1, 7, 13, 22, 36, 45]  # números escolhidos pelo jogador

num_quadras = 0  # contador de acertos da quina

for i in range(n):
    numeros_sorteados = []
    while len(numeros_sorteados) < 6:
        numero = random.randint(1, 60)
        if numero not in numeros_sorteados:
            numeros_sorteados.append(numero)
    numeros_sorteados.sort()

    numeros_certos = 0
    for numero in meu_bilhete:
        if numero in numeros_sorteados:
            numeros_certos += 1

    if numeros_certos == 4:
        num_quadras += 1

print("Probabilidade de acertar a quadra em um jogo:", num_quadras/n)

Probabilidade de acertar a quadra em um jogo: 0.000422


Questão 3: Considere uma disputa em que o jogador A jogue com a pontos e o jogador B com b pontos. Uma partida deste jogo consiste em 

1) o jogador A lança 2 dados e soma o valor das duas faces com a, 

2) o jogador B lança 2 dados e soma aos seus pontos b 

3) o jogador que obtiver o maior resultado final vence a rodada,

4) o vencedor do jogo é aquele que for o primeiro a vencer 10 rodadas. 

Com o auxílio de um programa de computador, estime a probabilidade de que o jogador A vença quando
1. a = 7 e b = 8.
2. a = 8 e b = 10.
3. a = 6 e b = 10.
4. a = 8 e b = 11.

In [10]:
import random

def lancar_dados():
    return random.randint(1, 6) + random.randint(1, 6)

def jogar_partida(pontos_a, pontos_b):

    vitorias_a = 0
    vitorias_b = 0
    while vitorias_a < 10 and vitorias_b < 10:
        resultado_a = lancar_dados() + pontos_a
        resultado_b = lancar_dados() + pontos_b
        if resultado_a > resultado_b:
            vitorias_a += 1
        elif resultado_b > resultado_a:
            vitorias_b += 1
    if vitorias_a == 10:
        return True
    elif vitorias_b == 10:
        return False

def estimar_probabilidade(num_partidas):
    vitorias_a = 0
    for i in range(num_partidas):
        if jogar_partida(7, 8):
            vitorias_a += 1
    return vitorias_a / num_partidas

print(estimar_probabilidade(10000))


0.1363


In [14]:
import random

def lancar_dados():
    return random.randint(1, 6) + random.randint(1, 6)

def jogar_partida(pontos_a, pontos_b):

    vitorias_a = 0
    vitorias_b = 0
    while vitorias_a < 10 and vitorias_b < 10:
        resultado_a = lancar_dados() + pontos_a
        resultado_b = lancar_dados() + pontos_b
        if resultado_a > resultado_b:
            vitorias_a += 1
        elif resultado_b > resultado_a:
            vitorias_b += 1
    if vitorias_a == 10:
        return True
    elif vitorias_b == 10:
        return False

def estimar_probabilidade(num_partidas):
    vitorias_a = 0
    for i in range(num_partidas):
        if jogar_partida(8, 10):
            vitorias_a += 1
    return vitorias_a / num_partidas

print(estimar_probabilidade(10000))

0.0126


In [16]:
import random

def lancar_dados():
    return random.randint(1, 6) + random.randint(1, 6)

def jogar_partida(pontos_a, pontos_b):

    vitorias_a = 0
    vitorias_b = 0
    while vitorias_a < 10 and vitorias_b < 10:
        resultado_a = lancar_dados() + pontos_a
        resultado_b = lancar_dados() + pontos_b
        if resultado_a > resultado_b:
            vitorias_a += 1
        elif resultado_b > resultado_a:
            vitorias_b += 1
    if vitorias_a == 10:
        return True
    elif vitorias_b == 10:
        return False

def estimar_probabilidade(num_partidas):
    vitorias_a = 0
    for i in range(num_partidas):
        if jogar_partida(6, 10):
            vitorias_a += 1
    return vitorias_a / num_partidas

print(estimar_probabilidade(10000))

0.0


In [18]:
import random

def lancar_dados():
    return random.randint(1, 6) + random.randint(1, 6)

def jogar_partida(pontos_a, pontos_b):

    vitorias_a = 0
    vitorias_b = 0
    while vitorias_a < 10 and vitorias_b < 10:
        resultado_a = lancar_dados() + pontos_a
        resultado_b = lancar_dados() + pontos_b
        if resultado_a > resultado_b:
            vitorias_a += 1
        elif resultado_b > resultado_a:
            vitorias_b += 1
    if vitorias_a == 10:
        return True
    elif vitorias_b == 10:
        return False

def estimar_probabilidade(num_partidas):
    vitorias_a = 0
    for i in range(num_partidas):
        if jogar_partida(8, 11):
            vitorias_a += 1
    return vitorias_a / num_partidas

print(estimar_probabilidade(10000))

0.0003
